File: commerce-service/src/main/java/com/singhtwenty2/ssew_core/controller/polling/PollingController.java
/**
 * Copyright 2025 SSEW Core Service
 * Developer: Aryan Singh (@singhtwenty2)
 * Portfolio: https://singhtwenty2.pages.dev/
 * This file is part of SSEW E-commerce Backend System
 * Licensed under MIT License
 * For commercial use and inquiries: aryansingh.corp@gmail.com
 * @author Aryan Singh (@singhtwenty2)
 * @project SSEW E-commerce Backend System
 * @since 2025
 */
package com.singhtwenty2.ssew_core.controller.polling;

import com.singhtwenty2.ssew_core.data.dto.common.GlobalApiResponse;
import com.singhtwenty2.ssew_core.data.dto.polling.PoolingDTO.PollEvent;
import com.singhtwenty2.ssew_core.data.dto.polling.PoolingDTO.PollResponse;
import com.singhtwenty2.ssew_core.data.enums.PollEventType;
import com.singhtwenty2.ssew_core.service.polling.RedisEventService;
import com.singhtwenty2.ssew_core.util.io.AuthenticationUtils;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/v1/polling")
@RequiredArgsConstructor
@Slf4j
public class PollingController {

    private final RedisEventService redisEventService;

    private static final long SHORT_POLL_INTERVAL_MS = 5000;
    private static final long LONG_POLL_INTERVAL_MS = 30000;

    @GetMapping("/events")
    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
    public ResponseEntity<GlobalApiResponse<PollResponse>> pollEvents(
            @RequestParam PollEventType eventType,
            @RequestParam(required = false) String lastEventId,
            @RequestParam(defaultValue = "false") boolean longPoll,
            Authentication authentication,
            HttpServletRequest request
    ) {
        String userId = AuthenticationUtils.extractUserId(authentication, request, "poll events");
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(
                    GlobalApiResponse.<PollResponse>builder()
                            .success(false)
                            .message("Unauthorized access")
                            .build()
            );
        }

        boolean isAdmin = authentication.getAuthorities().stream()
                .anyMatch(auth -> auth.getAuthority().equals("ROLE_ADMIN"));

        List<PollEvent> events = retrieveEvents(eventType, userId, lastEventId, isAdmin);

        if (longPoll && events.isEmpty()) {
            events = waitForEvents(eventType, userId, lastEventId, isAdmin);
        }

        PollResponse response = PollResponse.builder()
                .events(events)
                .lastEventId(events.isEmpty() ? lastEventId :
                        events.getLast().getEventId())
                .pollInterval(longPoll ? LONG_POLL_INTERVAL_MS : SHORT_POLL_INTERVAL_MS)
                .hasMore(events.size() >= 50)
                .build();

        return ResponseEntity.ok(
                GlobalApiResponse.<PollResponse>builder()
                        .success(true)
                        .message("Events retrieved")
                        .data(response)
                        .build()
        );
    }

    @GetMapping("/admin/events")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<GlobalApiResponse<PollResponse>> pollAdminEvents(
            @RequestParam(required = false) String lastEventId,
            @RequestParam(defaultValue = "false") boolean longPoll,
            HttpServletRequest request
    ) {
        log.debug("Admin polling events from IP: {}", request.getRemoteAddr());

        List<PollEvent> events = redisEventService.getAdminEvents(lastEventId);

        if (longPoll && events.isEmpty()) {
            events = waitForAdminEvents(lastEventId);
        }

        PollResponse response = PollResponse.builder()
                .events(events)
                .lastEventId(events.isEmpty() ? lastEventId :
                        events.getLast().getEventId())
                .pollInterval(longPoll ? LONG_POLL_INTERVAL_MS : SHORT_POLL_INTERVAL_MS)
                .hasMore(events.size() >= 50)
                .build();

        return ResponseEntity.ok(
                GlobalApiResponse.<PollResponse>builder()
                        .success(true)
                        .message("Admin events retrieved")
                        .data(response)
                        .build()
        );
    }

    @GetMapping("/user/events")
    @PreAuthorize("hasRole('USER') or hasRole('ADMIN')")
    public ResponseEntity<GlobalApiResponse<PollResponse>> pollUserEvents(
            @RequestParam(required = false) String lastEventId,
            @RequestParam(defaultValue = "false") boolean longPoll,
            Authentication authentication,
            HttpServletRequest request
    ) {
        String userId = AuthenticationUtils.extractUserId(authentication, request, "poll user events");
        if (userId == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(
                    GlobalApiResponse.<PollResponse>builder()
                            .success(false)
                            .message("Unauthorized access")
                            .build()
            );
        }

        List<PollEvent> events = redisEventService.getUserEvents(userId, lastEventId);

        if (longPoll && events.isEmpty()) {
            events = waitForUserEvents(userId, lastEventId);
        }

        PollResponse response = PollResponse.builder()
                .events(events)
                .lastEventId(events.isEmpty() ? lastEventId :
                        events.getLast().getEventId())
                .pollInterval(longPoll ? LONG_POLL_INTERVAL_MS : SHORT_POLL_INTERVAL_MS)
                .hasMore(events.size() >= 50)
                .build();

        return ResponseEntity.ok(
                GlobalApiResponse.<PollResponse>builder()
                        .success(true)
                        .message("User events retrieved")
                        .data(response)
                        .build()
        );
    }

    private List<PollEvent> retrieveEvents(PollEventType eventType, String userId,
                                           String lastEventId, boolean isAdmin) {
        if (isAdmin && (eventType == PollEventType.ADMIN_NEW_ORDER ||
                        eventType == PollEventType.ADMIN_ORDER_UPDATE)) {
            return redisEventService.getAdminEvents(lastEventId);
        } else {
            return redisEventService.getEventsByType(eventType, userId, lastEventId);
        }
    }

    private List<PollEvent> waitForEvents(PollEventType eventType, String userId,
                                          String lastEventId, boolean isAdmin) {
        long startTime = System.currentTimeMillis();
        long timeout = 25000;

        while (System.currentTimeMillis() - startTime < timeout) {
            List<PollEvent> events = retrieveEvents(eventType, userId, lastEventId, isAdmin);
            if (!events.isEmpty()) {
                return events;
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }

        return List.of();
    }

    private List<PollEvent> waitForAdminEvents(String lastEventId) {
        long startTime = System.currentTimeMillis();
        long timeout = 25000;

        while (System.currentTimeMillis() - startTime < timeout) {
            List<PollEvent> events = redisEventService.getAdminEvents(lastEventId);
            if (!events.isEmpty()) {
                return events;
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }

        return List.of();
    }

    private List<PollEvent> waitForUserEvents(String userId, String lastEventId) {
        long startTime = System.currentTimeMillis();
        long timeout = 25000;

        while (System.currentTimeMillis() - startTime < timeout) {
            List<PollEvent> events = redisEventService.getUserEvents(userId, lastEventId);
            if (!events.isEmpty()) {
                return events;
            }

            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                break;
            }
        }

        return List.of();
    }
}

File: telemetry-service/internal/controllers/polling_controller.go
// Copyright 2025 Aryan Singh
// Developer: Aryan Singh (@singhtwenty2)
// Portfolio: https://singhtwenty2.pages.dev/
// This file is part of SSEW E-commerce Backend System
// Licensed under MIT License
// For commercial use and inquiries: aryansingh.corp@gmail.com
// @author Aryan Singh (@singhtwenty2)
// @project SSEW E-commerce Backend System
// @since 2025

package controllers

import (
	"context"
	"net/http"
	"time"

	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/config"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/dto"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/enums"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/middleware"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/models"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/service"
	"github.com/callmeysg/SSEW-Backend/telemetry-service/internal/utils"
	"github.com/gin-gonic/gin"
)

type PollingController struct {
	eventService service.EventService
}

func NewPollingController(eventService service.EventService) *PollingController {
	return &PollingController{
		eventService: eventService,
	}
}

func (pc *PollingController) HealthCheck(c *gin.Context) {
	ctx, cancel := context.WithTimeout(c.Request.Context(), 2*time.Second)
	defer cancel()

	if err := pc.eventService.CheckHealth(ctx); err != nil {
		utils.ErrorResponse(c, http.StatusServiceUnavailable, "Service is unhealthy", map[string]interface{}{
			"service": "telemetry",
			"status":  "unhealthy",
			"dependencies": map[string]string{
				"redis": "disconnected",
			},
			"error": err.Error(),
		})
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "Telemetry service is healthy", map[string]interface{}{
		"service": "telemetry",
		"status":  "healthy",
		"dependencies": map[string]string{
			"redis": "connected",
		},
	})
}

func (pc *PollingController) PollEvents(c *gin.Context) {
	var req dto.PollRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request parameters")
		return
	}

	if req.EventType != "" && !enums.IsValidEventType(req.EventType) {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid event type")
		return
	}

	userId := middleware.GetUserID(c)
	if userId == "" {
		utils.ErrorResponse(c, http.StatusUnauthorized, "Unauthorized access")
		return
	}

	isAdmin := middleware.IsAdmin(c)

	ctx := c.Request.Context()
	var events []models.PollEvent
	var err error

	fetchFunc := func() ([]models.PollEvent, error) {
		if req.EventType != "" {
			return pc.eventService.GetEventsByType(ctx, enums.PollEventType(req.EventType), userId, req.LastEventID)
		}

		if isAdmin && (req.EventType == string(enums.AdminNewOrder) || req.EventType == string(enums.AdminOrderUpdate)) {
			return pc.eventService.GetAdminEvents(ctx, req.LastEventID)
		}

		return pc.eventService.GetUserEvents(ctx, userId, req.LastEventID)
	}

	if req.LongPoll {
		events, err = pc.eventService.PollEventsWithTimeout(ctx, fetchFunc, config.AppConfig.LongPollTimeout)
	} else {
		events, err = fetchFunc()
	}

	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to retrieve events")
		return
	}

	response := pc.eventService.BuildPollResponse(events, req.LastEventID, req.LongPoll)
	utils.SuccessResponse(c, http.StatusOK, "Events retrieved", response)
}

func (pc *PollingController) PollAdminEvents(c *gin.Context) {
	var req dto.AdminPollRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request parameters")
		return
	}

	ctx := c.Request.Context()
	var events []models.PollEvent
	var err error

	fetchFunc := func() ([]models.PollEvent, error) {
		return pc.eventService.GetAdminEvents(ctx, req.LastEventID)
	}

	if req.LongPoll {
		events, err = pc.eventService.PollEventsWithTimeout(ctx, fetchFunc, config.AppConfig.LongPollTimeout)
	} else {
		events, err = fetchFunc()
	}

	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to retrieve events")
		return
	}

	response := pc.eventService.BuildPollResponse(events, req.LastEventID, req.LongPoll)
	utils.SuccessResponse(c, http.StatusOK, "Admin events retrieved", response)
}

func (pc *PollingController) PollUserEvents(c *gin.Context) {
	var req dto.UserPollRequest
	if err := c.ShouldBindQuery(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request parameters")
		return
	}

	userId := middleware.GetUserID(c)
	if userId == "" {
		utils.ErrorResponse(c, http.StatusUnauthorized, "Unauthorized access")
		return
	}

	ctx := c.Request.Context()
	var events []models.PollEvent
	var err error

	fetchFunc := func() ([]models.PollEvent, error) {
		return pc.eventService.GetUserEvents(ctx, userId, req.LastEventID)
	}

	if req.LongPoll {
		events, err = pc.eventService.PollEventsWithTimeout(ctx, fetchFunc, config.AppConfig.LongPollTimeout)
	} else {
		events, err = fetchFunc()
	}

	if err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to retrieve events")
		return
	}

	response := pc.eventService.BuildPollResponse(events, req.LastEventID, req.LongPoll)
	utils.SuccessResponse(c, http.StatusOK, "User events retrieved", response)
}

func (pc *PollingController) PublishEvent(c *gin.Context) {
	var req models.EventPublishRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request body")
		return
	}

	if !enums.IsValidEventType(string(req.EventType)) || !enums.IsValidActionType(string(req.Action)) {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid event type or action")
		return
	}

	ctx := context.Background()
	if err := pc.eventService.PublishGenericEvent(ctx, &req); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to publish event")
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "Event published successfully", nil)
}

func (pc *PollingController) PublishOrderStatusChange(c *gin.Context) {
	var req models.OrderStatusChangeRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request body")
		return
	}

	ctx := context.Background()
	if err := pc.eventService.PublishOrderStatusChangeEvent(ctx, req.OrderID, req.UserID, req.NewStatus); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to publish order status change event")
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "Order status change event published", nil)
}

func (pc *PollingController) PublishNewOrderEvent(c *gin.Context) {
	var req models.NewOrderEventRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request body")
		return
	}

	ctx := context.Background()
	if err := pc.eventService.PublishNewOrderEventForAdmin(ctx, req.OrderID, req.CustomerName, req.TotalAmount); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to publish new order event")
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "New order event published", nil)
}

func (pc *PollingController) PublishOrderUpdateEvent(c *gin.Context) {
	var req models.OrderUpdateEventRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.ErrorResponse(c, http.StatusBadRequest, "Invalid request body")
		return
	}

	ctx := context.Background()
	if err := pc.eventService.PublishOrderUpdateEventForAdmin(ctx, req.OrderID, req.UpdateType, req.Details); err != nil {
		utils.ErrorResponse(c, http.StatusInternalServerError, "Failed to publish order update event")
		return
	}

	utils.SuccessResponse(c, http.StatusOK, "Order update event published", nil)
}
